### ISSUE TITLE
Problem trying to run evaluate with downloaded checkpoint

### ISSUE THREAD

**Issue by JonaSchulz** [nicklashansen/tdmpc2#52]
I'm trying to run evaluate.py with a downloaded checkpoint model (dog-run-1.pt). However, I get the following error:

"You are trying to pop key `('__batch_size',)` which is not in dict without providing default value. Keys=_TensorDictKeysView(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'],\n    include_nested=True,\n    leaves_only=False)."
KeyError: 'key "__batch_size" not found in TensorDict with keys [\'0\', \'1\', \'2\', \'3\', \'4\', \'5\', \'6\', \'7\', \'8\', \'9\']'

The above exception was the direct cause of the following exception:

  File "/home/idsc/tdmpc2/tdmpc2/tdmpc2.py", line 86, in load
    self.model.load_state_dict(state_dict["model"])
  File "/home/idsc/tdmpc2/tdmpc2/evaluate.py", line 59, in evaluate
    agent.load(cfg.checkpoint)
  File "/home/idsc/tdmpc2/tdmpc2/evaluate.py", line 103, in <module>
    evaluate()
KeyError: "You are trying to pop key `('__batch_size',)` which is not in dict without providing default value. Keys=_TensorDictKeysView(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'],\n    include_nested=True,\n    leaves_only=False)."

I didn't make any modifications to the code and use all default parameters from the config.yaml file

**Comment by nicklashansen** [comment]
Hi @JonaSchulz, I believe this error might be related to recent efforts to speed up the codebase. One temporary solution would be to go back to an earlier commit, e.g. https://github.com/nicklashansen/tdmpc2/commit/88095e7899497cf7a1da36fb6bbb6bc7b5370d53, while I work on issuing a fix for this.

**Comment by nicklashansen** [comment]
@vmoens Hi, if you have a moment at some point, would you mind providing some guidance re: this issue? It appears that our recent changes to the Q-function module breaks loading of older checkpoints. All other parts of the model seem to load correctly, and I haven't quite been able to find a solution but the error is coming from `tensordict`. See commit https://github.com/nicklashansen/tdmpc2/commit/2e27fbb6f4d026520de1a2d5a8c4a0e9c2d076f9 for a breakdown of the problem.

**Comment by vmoens** [comment]
Do you have a minimal repro? I can fix that easily but I need to make sure I'm loading things as expected

**Comment by nicklashansen** [comment]
Yes! You can download a checkpoint from here https://huggingface.co/nicklashansen/tdmpc2/tree/main/dmcontrol e.g. `wget https://huggingface.co/nicklashansen/tdmpc2/resolve/main/dmcontrol/dog-run-1.pt?download=true` and then run `python evaluate.py task=dog-run checkpoint="path/to/checkpoint.pt"`. Without the try-catch this will produce the error that @JonaSchulz posted here, and with the changes I made in commit https://github.com/nicklashansen/tdmpc2/commit/2e27fbb6f4d026520de1a2d5a8c4a0e9c2d076f9 the code will successfully load all modules except the Q-functions and then raise a `NotImplementedError`.

**Comment by vmoens** [comment]
What's the most useful way for me to help (beyond the BC-fix in tensordict)?
Do you want a custom func to check/transform the state-dict, ie, a patch in your code base while waiting for a new tensordict release?

**Comment by nicklashansen** [comment]
I can't quite tell what is causing this error or how to fix it. I assumed that the error is because the checkpoint was generated last year using our previous Q-function implementation, and that we are trying to load them using the new Q-function implementation introduced with the `torch.compile` functionality. So a patch to this codebase that makes the appropriate conversion might be sufficient? I guess I'm just looking for pointers on how to actually do that now that we use tensordicts for the parameters.

**Comment by vmoens** [comment]
Ok then the work plan should be to write a pre hook for load state dict that maps the old format to the new.
I made some progress I'll keep you posted (at NeurIPS now with limited coding time / day)

**Comment by nicklashansen** [comment]
No problem! Take your time. Enjoy Vancouver!

**Comment by nicklashansen** [comment]
@vmoens Happy new year! Sorry for bringing this up again -- did you manage to find a solution or would you be able to point me in the right direction? I could give it a try myself

**Comment by vmoens** [comment]
See #55 for a potential solution

**Comment by MihaiStanusoiu** [comment]
Would it be better to save the uncompiled model, as it shares the weights with the compiled instance? This issue would then be avoided (although the conversion is still needed for already saved compiled models)

**Comment by nicklashansen** [comment]
Hi everyone, I finally have a complete fix to this issue. Commit https://github.com/nicklashansen/tdmpc2/commit/5ced6dfeb47ea51da1246d05471ec7cac7692a3c adds a `api_model_conversion` method that auto-checks whether a checkpoint loaded from disk was generated with our old model format and, if yes, converts it to the new format. I have checked that converted models perform as expected. Thanks everyone for your patience!

**Comment by vmoens** [comment]
Thanks @nicklashansen 
I will make this API easier to use for the next projects!
